ipsec: Implement RPS for ESP packets

learnitall opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [X] Thanks for contributing!

This PR implements Receive Packet Steering (RPS) for ESP packets to address the single-core processing limitation that can negatively impact high-throughput workloads. Benchmarks in an Equinix Metal Kubernetes cluster showed throughput increases by up to 430%.

### The Problem

ESP packets for a single IPSec tunnel are not subject to RSS on ingress into a node. An IPSec tunnel hides the underlying flows, therefore any RSS implementation will view the tunnel as a single flow and route all ESP packets to a single CPU for processing. This hinders throughput by forcing all packets in an IPSec tunnel to be processed sequentially.

### This PR

This PR implements support for RSS with ESP packets by stuffing the inner packet's skb hash into the SPI field. Cilium uses a limited number of bits within the SPI, meaning there is extra space to be safely utilized. After a packet has been sent through the XFRM stack for encryption and ends up back in `bpf_host.c`, the hash stored in the skb is added into the lower three bytes of the SPI field. The last byte is reserved for the actual SPI. When an ESP packet is received by a node, `bpf_xdp.c` pulls out the stuffed hash, restoring the SPI field in the ESP packet to its true value, and uses the hash to make a CPU routing decision. This allows for flows within an IPSec tunnel to be deterministically spread across CPUs, increasing throughput.

This implementation leverages the kernel's lazy evaluation of the packet hash. It assumes that the hash in the skb of an ESP packet is the hash of the inner, encrypted packet. Since Cilium's datapath does not perform hash invalidation, since the use case for doing so has not presented itself, the hash shouldn't change after a packet is encrypted by the XFRM stack.

To see this in action, create an IPSec tunnel and use the following bpftrace script to observe packets entering `ip_output` and `xfrm_output`. You'll see cases where a packet enters `xfrm_output` with a certain hash and afterwards an ESP packet enters `ip_output` with the same hash:

```
#include<linux/ip.h>
kfunc:vmlinux:xfrm_output
{
  $data = args->skb->head + args->skb->network_header;
  $ip_header = (struct iphdr *)$data;
  printf("{\"source\":\"%s\",\"proto\":\"%d\",\"src\":\"0x%x\",\"dst\":\"0x%x\",\"hash\":\"0x%x\"}\n", "xfrm_output", $ip_header->protocol, $ip_header->saddr, $ip_header->daddr, args->skb->hash);

}

kfunc:vmlinux:ip_output
{
  $data = args->skb->head + args->skb->network_header;
  $ip_header = (struct iphdr *)$data;
  printf("{\"source\":\"%s\",\"proto\":\"%d\",\"src\":\"0x%x\",\"dst\":\"0x%x\",\"hash\":\"0x%x\"}\n", "ip_output", $ip_header->protocol, $ip_header->saddr, $ip_header->daddr, args->skb->hash);
}
```

### Benchmarks

I tested this on a Kubernetes cluster deployed on three (one CP, two workers) m3.small.x86 (eight cores hyperthreaded, 25Gb/s NIC) Equinix Metal machines. I ran netperf stream, RR and CRR tests for Cilium configured with no encryption, standard IPSec encryption, WireGuard and IPSec w/ RPS. The results show the RPS solution successfully loadbalancing flows in an IPSec tunnel across CPUs.

![rps-stream](https://github.com/user-attachments/assets/9347b327-c795-4d31-9a74-be63187d4e3f)

Stream performance shows the throughput scaling with the number of clients. With 16 clients, the RPS implementation in this PR provides a 430% increase in throughput over standard IPSec and a 289% throughput increase over WireGuard. A follow-up to this PR will involve investigation into why the linear scaling of throughput stops about halfway though.

![tcp_rr_p99](https://github.com/user-attachments/assets/739684a1-9060-4ab8-a121-b39ae35958b8)

The p99 RR results shows that the RPS implementation does not negatively impact request-response workloads, event with the CPU redirection in the hot-path.


![tcp_crr_p99](https://github.com/user-attachments/assets/818c9837-5e52-4bde-9bc7-e5a12b8d63a1)


The p99 CRR results show that the RPS implementation may increase latency for CRR-type workloads with a small number of clients, but the overhead of the implementation is canceled out by performance gains as the number of clients increases.

The implementation for the benchmark automation can be found at https://github.com/cilium/scaffolding/pull/148.

```release-note
Add experimental feature to increase IPSec throughput through Receive Packet Steering (RPS)
```

EDIT: I presented this at KubeCon and made a spreadsheet with all the testing data and some nicer graphs. It can be found here: https://docs.google.com/spreadsheets/d/16BxVG9OEXpqEBFDEwmjvcDUqP0RTD1W6Hb9EgADRBIM/edit?usp=sharing.

learnitall committed:
xdp: Add TestCell

This commit adds a test cell to the XDP configuration package to allow
for unit tests to consume XDP configuration. This is useful for features
which depend on XDP configuration and want to validate in unit tests
that their logic for handling XDP configuration is sound.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
xdp: Add WithEnforceXDPNative validator

This commit adds a validator to the XDP configuration package to ensure
that the materialized XDP mode is "native", returning an error during
Hive startup otherwise.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ipsecrps [1/3]: Add configuration cell for IPSec RPS mode

This commit introduces a new configuratin cell for a newly added IPSec
feature: Received Packet Steering. The cell added in this commit allows
for users to enable the feature and performs configuration validation,
ensuring that when the feature is enabled, IPSec and XDP Native mode are
also enabled.

The next commit contains the datapath implementation.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
xdp: Fix flake in XDP config validation

This commit fixes a bug causing flaky behavior in the handling of XDP
Validators. Previously, a validator from an enabler was only executed if
the mode the enabler is requesting is different than the currently set
mode. Since enablers are iterated over in a for loop, this caused random
failures based on the order enablers were evaluated in. To fix this,
let's unconditionally execute all validators from all passed in
enablers.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
maps: Add CPUMap for XDP redirection

This commit adds a new map, named CPUMap, which a BPF_MAP_TYPE_CPUMAP
that can be used by XDP programs for packet steering. The map will only
be enabled if requested by another cell and will automatically populate
one entry for each CPU detected on the local system.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
xdp: Various quality fixes in XDP config cell tests

This commit fixes various quality issues in the XDP config cell unit
tests:

* Add a missing check in the XDP Config test on the error
  returned by hive population, ensuring that no error is returned
  when one wasn't expected.
* Organize members in the test case struct for memory efficiency.
* Replace calls to `Error` followed by `FailNow` with `Fatal`

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ipsecrps [2/3]: Implement packet hash SPI stashing

This commit implements Receive Packet Steering (RPS) for flows within an
IPSec tunnel. This is implemented by stashing inner packet hash's within
the SPI field of egressing ESP packets. This hash is removed on ingress
by an XDP program which then performs packet steering to available CPUs.

Netperf TCP_STREAM benchmarks in an Equinix Metal Kubernetes cluster
showed throguhput gains by up to 430% when the number of netperf clients
matched the number of cores on each node.

The next commit adds the feature to the documentation.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
docs, ipsecrps [3/3]: Document IPSec XDP Acceleration

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

gandro commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1012,6 +1012,8 @@ encryption:
     keyRotationDuration: "5m"
     # -- Enable IPsec encrypted overlay
     encryptedOverlay: false
+    # -- Enable Receive Packet Steering (RPS) for ESP packets.
```
If this is experimental, I think we should call this out here, e.g. 
```suggestion
    # -- Enable Receive Packet Steering (RPS) for ESP packets (experimental).
```

learnitall replied:
Good idea, thanks!

gandro commented on pkg/maps/cpumap/map.go:
```diff
@@ -0,0 +1,161 @@
+// Populate should be called on hive startup to populate the CPUMap
+// with entries for each CPU.
+func (c *cpuMap) Populate() error {
+	slog.Warn("populating cpumap")
```
Stray debug comment?

learnitall replied:
Yes thank you, there were a couple in there :facepalm:

gandro commented on pkg/maps/cpumap/config.go:
```diff
@@ -0,0 +1,114 @@
+// If the cpumap is enabled or not.
+func (c Config) Enabled() bool { return c.enabled }
+
+// The number of CPUs that are enabled on the node.
+func (c Config) NumCPUs() uint { return c.numCPUs }
```
Technically this is the number of possible CPUs on the node, not the number of enabled CPUs on the node, right? In particular, I'm reminded of issues we had in the past when the two values were confused: https://github.com/cilium/cilium/pull/12121

learnitall replied:
Ah good point. TIL about CPU hot-swapping. Thanks!

gandro replied:
Since I don't see any changes, are we certain that `runtime.NumCPU()` gives us the correct value here, and that we shouldn't use `common.GetNumPossibleCPUs` which potentially returns a higher value? 

All other per-CPU BPF maps use `common.GetNumPossibleCPUs` after all. The `__NR_CPUS__` datapath equivalent is initialized with `common.GetNumPossibleCPUs()`, i.e. the number of possible CPUs. Using `NumCPU()` in the agent means the datapath and the agent are not agreeing on the value of `__NR_CPUS__`.

learnitall replied:
So this value is set [here](https://github.com/cilium/cilium/pull/35661/files#diff-e3c20aac539bc98432fb973e781277d771069860a2fda58fa0857fdaf4ec8201R104). It used to use `GetNumPossibleCPUs` and if an error was returned, fall back to `runtime.NumCPU`. I changed it to just `NumCPU`.

I'm 90% sure that the map must be set with the number of actual CPUs on the node rather than the number of possible CPUs. The map is used to perform packet steering from one CPU to another and assumes that each entry in the map is a valid redirection target, therefore if the map size is not equal to the number of CPUs available to perform redirection we'll start seeing packet drops.

Thanks for pointing out that `__NR_CPUS__` is set to the number of possible CPUs and not the number of actual CPUs. TIL. I'll need to add a new macro which plumbs the actual number of CPUs into the datapath for this map.

gandro replied:
Thanks for confirming! Yeah I think it makes sense then to introduce a new macro for the datapath. 

It might also make sense to think about what would happen if more CPUs are attached (this is probably fine, we're just not using them), but also what happens if CPUs are removed. To be honest, I'm not even sure if the rest of Cilium deals gracefully with this.

learnitall replied:
Hm that's a good point. If CPUs are removed then there would be packet drops, since the XDP program would try to redirect to a CPU that isn't there anymore. I could add a warning in the docs about this to make it clear. Maybe we could come up with a better solution as a follow-up?

gandro replied:
Yes, I absolutely think this is something to tackle in a follow-up. As I've said, I'm not even sure if the rest of Cilium deals with this correctly either. So this would need some experimentation for sure.

learnitall replied:
Ok cool, sounds like a plan! I added a new macro which contains the number of online CPUs on the system.

gandro commented on pkg/datapath/linux/ipsecrps/ipsec_rps_linux.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+func (u userFlags) Flags(flags *pflag.FlagSet) {
+	flags.Bool("enable-ipsec-acceleration", u.EnableIpsecAcceleration, "Enable acceleration for IPSec using RPS.")
```
Similarly to above, probably worth adding an `(experimental)` into the description here so users don't turn this on without consulting the docs

learnitall replied:
Done!

gandro reviewed (commented):
Awesome work! I'm not an IPSec or BPF expert, so I only did I did a cursory review of the agent parts so far

nathanjsweet commented on pkg/datapath/xdp/xdp_test.go:
```diff
@@ -117,6 +117,9 @@ func TestConf(t *testing.T) {
 
 				return
+			} else if err != nil {
+				t.Errorf("unexpected error from hive: %s", err.Error())
```
Please convert this to :
```golang
t.Fatalf("unexpected error from hive: %s", err.Error())
```
and remove the `t.FailNow()`

learnitall replied:
Ah nice, thanks!

nathanjsweet reviewed (approved):
LGTM.

pchaigno commented on pkg/maps/cpumap/config.go:
```diff
@@ -0,0 +1,111 @@
+var defaultUserConfig = userCfg{
+	// TODO: This is picked based off of some initial testing, but
+	// it should probably be tuned based off of the system's specs.
+	XdpCpumapQSize: 255,
```
So should we have an agent flag to tune this? It can be a follow-up of course, but I'd rather keep track of it in an issue than as a TODO. We tend to forget about TODOs more easily.

learnitall replied:
The flag `--xdp-cpumap-qsize` is included to make this configurable, but it's not exposed via helm. Were you thinking it should be? I'll make an issue to replace the TODO though, that's a good idea. Thanks!

pchaigno replied:
I didn't notice the related flag :sweat_smile: But if we have the flag, why do we need the TODO? Are you suggesting we should tune the default based on the user's system? That feels like overengineering. If users care enough about performance to enable this feature, I guess they'll be okay tuning their system to get an extra couple pourcents, no?

learnitall replied:
> Are you suggesting we should tune the default based on the user's system? That feels like overengineering.

Hm ok, I see what you're saying. I think that the TODO item then can be centered more around doing some testing to see if there is a better default out there, and based on that testing, document any potential recommendations. I picked 255 fairly arbitrarily and it just happened to give some good results, so doing some testing to see if there is a better option would be great.

pchaigno commented on pkg/maps/cpumap/config.go:
```diff
@@ -0,0 +1,111 @@
+
+	// TODO: Just picking a sensible default for now that worked well in testing, but
+	// this should by dynamic based on the available system memory.
+	cfg.qsize = in.UserCfg.XdpCpumapQSize
```
Same concern as for the other code TODO :)

pchaigno commented on pkg/datapath/linux/ipsecrps/ipsec_rps_linux.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+func (u userFlags) Flags(flags *pflag.FlagSet) {
+	flags.Bool("enable-ipsec-acceleration", u.EnableIpsecAcceleration, "Enable acceleration for IPSec using RPS (experimental).")
```
I agree with @gandro that we need to properly label this, but I think `beta` would be more appropriate. We already have a notion of "beta features" and there's precedent, whereas I don't think we have "experimental features". Users may be confused as to what the difference is.

learnitall replied:
Ah that makes sense, good idea. I'll change it up.

gandro commented on pkg/datapath/linux/ipsecrps/ipsec_rps_linux.go:
```diff
@@ -0,0 +1,147 @@
+
+	if cfg.Enabled() {
+		defines["ENABLE_IPSEC_RPS"] = "1"
+		defines["IPSEC_RPS_NCPU"] = strconv.FormatInt(int64(runtime.NumCPU()), 10)
```
Similar concerns to above, why `runtime.NumCPU()` and not `common.GetNumPossibleCPUs()`?

learnitall replied:
Oh hey that's technically dead code which I used before I learned about `__NR_CPUS__`, thank you! Please see https://github.com/cilium/cilium/pull/35661#discussion_r1856860866 for why I think this should stick around though and replace `__NR_CPUS__`.

learnitall committed:
datapath: Add __NR_ONLINE_CPUS__ macro variable

This commit adds a new variable to the datapath named
`__NR_ONLINE_CPUS__`, which is a count of the number of online
CPUs on the current node. This is distinct from the number of
possible CPUs exposed via __NR_CPUS__, which indicates the number
of CPUs the kernel is able to support (via hot-plugging).

The `tklauser/numcpus` library is used to determine the number of
online CPUs, as it is written by a core Cilium developer, handles
multiple platforms, and implements the exact functionality required.

This macro will be used in subsequent commits to build a CPUMAP
for XDP redirection.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

gandro reviewed (approved):
Awesome, looks good to me from an agent and "num CPU" perspective!

marseel commented on vendor/github.com/tklauser/numcpus/numcpus_linux.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+func getOnline() (int, error) {
+	if n, err := getFromCPUAffinity(); err == nil {
```
I'm not a kernel expert, but it seems it can also return not all online CPUs, but CPUs that are set by affinity for Cilium process?

pchaigno commented on bpf/Makefile:
```diff
@@ -134,7 +135,8 @@ XDP_OPTIONS = $(LB_OPTIONS) \
 	-DDISABLE_LOOPBACK_LB:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DENABLE_CAPTURE:-DDSR_ENCAP_MODE:-DDSR_ENCAP_IPIP:-DENABLE_SCTP:-DDSR_ENCAP_NONE=2 \
-	-DDISABLE_LOOPBACK_LB:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DENABLE_CAPTURE:-DDSR_ENCAP_MODE:-DDSR_ENCAP_GENEVE:-DENABLE_SCTP:-DDSR_ENCAP_IPIP=2
+	-DDISABLE_LOOPBACK_LB:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DENABLE_CAPTURE:-DDSR_ENCAP_MODE:-DDSR_ENCAP_GENEVE:-DENABLE_SCTP:-DDSR_ENCAP_IPIP=2 \
+	-DDISABLE_LOOPBACK_LB:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_IPSEC:-DENABLE_IPSEC_RPS:-DENABLE_CPU_MAP:-DFROM_HOST: \
```
Could you also add this macro to the configs in `bpf/complexity-tests/`? Those are used to generate the largest BPF programs possible and check they still pass the verifier.

learnitall replied:
Can do! Do you have guidance on how you'd like to see this done? We should be able to enable this on each kernel version but I'm not sure which config file(s) under each version to add the macro to.

pchaigno replied:
For a given program and a given kernel version, we typically have several files because some features are incompatible between them. Because of that, they are sort of local maximums for the complexity if you will (in the N+1 dimensional space where each of the N features is a dimension). So each file is trying to be one of the local maximums. More concretely, if ex. featA and featB are incompatible, we'd have one set of files maximizing complexity for featA and another for featB.

All that to say, in your case, AFAICS, the feature is not incompatible with anything but requires IPsec, so I'd add it to all files that have ENABLE_IPSEC, for the BPF programs that matter here.

learnitall replied:
Oh ok that makes a ton of sense, thanks for the explanation! I'll add it to to places where we have IPsec, but the feature is also incompatible with tunnel mode routing (if I remember correctly, when we enable tunnel mode routing and IPsec, then try to load the XDP program it leads to an import of an skb-only library), so I'll avoid places where that's enabled too.

pchaigno commented on bpf/bpf_xdp.c:
```diff
@@ -215,6 +220,28 @@ static __always_inline int check_v4_lb(struct __ctx_buff *ctx __maybe_unused)
+	cpu_dest = hash % __NR_CPUS__;
+
+	if (cpu_dest >= __NR_CPUS__)
+		return CTX_ACT_DROP;
```
Why is this check needed?

learnitall replied:
I saw this in the documentation for the CPU map [here](https://docs.kernel.org/bpf/map_cpumap.html), but you're right in that it shouldn't be needed. I'll drop it.

pchaigno commented on bpf/lib/ipsecrps.h:
```diff
@@ -0,0 +1,8 @@
+
+/* Masks are in network byte order */
+#define IPSEC_RPS_SPI_MASK  0xFF000000
+#define IPSEC_RPS_HASH_MASK 0x00FFFFFF
```
(This is going to be the painful comment from my review, sorry. I should have commented earlier on the CFP, but I didn't know what I know today about XFRM.)

What would be the impact of limiting the hash to only 8 bits? Could we make it configurable? If we do, what would be the impact of changing it?

The reason I'm asking is that we are currently misusing the SPI. I'm not even talking about your misuse, but what we already do in `main`. That is, it's recommended for each XFRM state to have its own SPI number. There doesn't seem to be any security consequence to that, but there is a performance consequence on Linux at scale (cf. https://pchaigno.github.io/cilium/2024/10/30/linux-xfrm-ipsec-reference-guide.html#data-structure-for-xfrm-states).

So at some point, we may likely want to push the node ID (16-bits) into the SPI field. Assuming we do that without changing the size of the key reference (8-bits), we will only have 8 bits left to encode the flow hash :grimacing:

learnitall replied:
I was wondering about this :smile_cat:. This would make a fairly significant difference from a theoretical standpoint. I ran a simulation to investigate, where for every possible 24-bit and 8-bit hash, and for all CPU counts in the range of 1-255, I calculated the target CPU that a flow would be steered to based on its hash and kept a tally of how many flows a CPU would receive. The count was weighted based on the number of 32-bit hashes that the smaller hashes represent. For instance, if a CPU would receive a flow for the 8-bit hash 0xFF, then it would also receive flows for all the 32-bit hashes in the range 0x000000FF-0xFFFFFFFF. This allows us to simulate the distribution of flows across CPUs on a node and get a sense of how even or uneven it is, based on using 8-bit or 24-bit hashes.

The results were pretty stark and show that only using 8-bits could have a fairly large impact. This plot has the number of CPUs on the x-axis and the "variance-to-mean ratio" (mean-normalized variance) on the y-axis. The red bars indicate cases where the y-axis has a value of zero:

![Figure 12](https://github.com/user-attachments/assets/118a4069-9a05-40c1-a151-ba7af777f5ff)

The scale of the y-axis for the 8-bit hashes (purple) and 24-bit hashes (blue) reveals all. The mean-normalized variances for the 8-bit hashes are in the millions, while the mean-normalized variances for the 24-bit hashes are less than one. In cases where the number of CPUs on a node isn't optimal (in-between a power of 2), the distributions of flows is tolerable for 24-bit hashes and pretty bad for 8-bit hashes.

This graph is a plot of the difference between the most amount of flows a CPU received and the smallest number of flows a CPU received, for a given number of bits in the hash and the given number of simulated CPUs. The values aren't normalized, so it becomes clear to tell that, with 8-bits, there's an extra 15 million possible flows that a CPU could receive, and with 24-bits, there's only an extra 250.

![Figure 11](https://github.com/user-attachments/assets/839e6344-268f-4f27-b25e-e227f0cd6a43)

So, **TL;DR: using 8-bit hashes should be fine as long as you have a number of CPUs equal to a power of 2, otherwise it's up in the are and entirely dependent on your workload.** I'll run another performance benchmark to see what happens with 8-bits in a more practical sense though.

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -359,6 +359,112 @@ Disabling Encryption
+IPsec tunnel to be processed in parallel. Spare room in the SPI field of each ESP packet is
+utilized to transfer the hash of the inner encrypted packet. An XDP program running on
+each node extracts the inner packet hash and uses it to perform deterministic packet
+steering across available CPU cores on the node.
```
The last two sentences here are too much information for a user guide IMO and can be removed.

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
This is IMO too strong. Reading this gives the impression traffic isn't encrypted anymore :sweat_smile: The flow hash allows attackers to distinguish the multiple flows in an IPsec tunnel, but it doesn't allow them to characterize the plain-text packets beyond that, does it?

How about something like:
> To enable load balancing of encrypted packet flows to cores, this feature encodes a flow hash of the plain-text traffic as part of the plain-text header of the encrypted traffic. An attacker will therefore be able to distinguish flows within the same IPsec tunnel. Ensure this is an acceptable risk before enabling this feature.

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
I'd remove the whole validation section. This is going into implementation details (what the SPI should look like) and I expect users can check the feature is working by checking their CPU usage.

qmonnet replied:
It could be some helpful info to move to the troubleshooting page, maybe

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
:+1: 

Another thing users will be worried about is the potential connectivity impact when enabling. AFAICS, if node A switched to RPS already, but node B didn't, then node A is going to send ESP traffic with an SPI node B cannot match against any XFRM state and traffic will be dropped for a while. Am I missing something?

pchaigno reviewed (changes requested):
Sorry for the delay :disappointed:

gandro commented on bpf/bpf_xdp.c:
```diff
@@ -215,6 +220,28 @@ static __always_inline int check_v4_lb(struct __ctx_buff *ctx __maybe_unused)
+
+	hash = esp->spi & IPSEC_RPS_HASH_MASK;
+	esp->spi = esp->spi & IPSEC_RPS_SPI_MASK;
+	cpu_dest = hash % __NR_CPUS__;
```
I missed that - shouldn't this be `__NR_ONLINE_CPUS__`?

learnitall replied:
:facepalm: yes it definitely should, thank you for catching this.

ti-mo replied:
I'm not aware of the inner workings of cpumap, but how does this prevent us from redirecting packets to a potentially-offline CPU? The 'online cpu count' is a little misleading because there can be holes in the range (see `/sys/devices/system/cpu/online` documentation) and notably the list of CPUs isn't 'defragmented' when changes happen. (e.g. CPU 0 and 1 can go offline)

learnitall replied:
That's a great point, I didn't think about the fragmentation aspect.

qmonnet commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -359,6 +359,112 @@ Disabling Encryption
+Decryption with Cilium IPsec is limited to a single CPU core per IPsec tunnel, which
+may impact performance in high-throughput workloads. An experimental flag can be
+given to the Cilium Agent to enable Receive Packet Steering (RPS) for IPsec ESP
+packets, which may improve throughput performance:
```
```suggestion
packets, which may improve throughput performance.
```

qmonnet reviewed (approved):
Reviewing doc changes: I agree with Paul's comments, but it looks good to me otherwise.

ti-mo commented on pkg/datapath/loader/compile.go:
```diff
@@ -100,15 +101,17 @@ type directoryInfo struct {
 
 var (
-	standardCFlags = []string{"-O2", "--target=bpf", "-std=gnu89",
+	standardCFlags = []string{
```
This looks like an arbitrary style change without a clear benefit.

ti-mo commented on pkg/datapath/loader/compile.go:
```diff
@@ -189,6 +197,7 @@ func compile(ctx context.Context, prog *progInfo, dir *directoryInfo) (string, e
 
 	compileArgs = append(compileArgs, standardCFlags...)
 	compileArgs = append(compileArgs, fmt.Sprintf("-D__NR_CPUS__=%d", possibleCPUs))
+	compileArgs = append(compileArgs, fmt.Sprintf("-D__NR_ONLINE_CPUS__=%d", onlineCPUs))
```
Adding an extra compile flag for me here is a no go for a few reasons:
- It's XDP-only AFAICS, and providing the define ambiently here means it's dead code in the majority of objects.
- I know it doesn't leave the preprocessor, but still. Someone (aka me) will need to hunt these down at some point and turn them into runtime-provided values to work towards the clang-free effort.
- Putting it in the preprocessor means this becomes harder to make reactive/dynamic to changes in the environment, e.g. CPUs going on/offline. A recompilation will always be necessary.
- This is one of a handful of ways to plumb config data into the datapath and we want to reduce this all down to one: `DECLARE_CONFIG`

I have more comments elsewhere, but if the variable turns out to be really needed, I suggest the following:

- `DECLARE_CONFIG(__u16, online_cpus, "Number of online CPUs for Receive Packet Steering")` at the top of `bpf_xdp.c`
- In `compileAndLoadXDPProg`,  in `CollectionOptions.Constants`, set the `online_cpus` key to the desired value

ti-mo commented on pkg/ebpf/map.go:
I think making `cilium_cpu_map` a singleton agent-managed map is a mistake and way overengineering things. 

Unless it needs live updates (your comment at the top of `cpumap/map.go` alludes to this not being the case due to not being supported), you can simply populate qlen in all keys right after loading the XDP program, see my comment above. If a use case arises later, we can introduce a reconciler, but given the kernel still lacks support, this may be a while.

Then you can drop the whole cpumap cell and pkg/cpumap.

learnitall replied:
> you can simply populate qlen in all keys right after loading the XDP program

Oh that's much simpler, thank you. I went with the cell because that was the simplest method I could think of to populate the map.

ti-mo commented on bpf/lib/maps.h:
```diff
@@ -284,3 +285,13 @@ tail_call_internal(struct __ctx_buff *ctx, const __u32 index, __s8 *ext_err)
+    __type(value, struct bpf_cpumap_val);
+    __uint(pinning, LIBBPF_PIN_BY_NAME);
+    __uint(max_entries, __NR_ONLINE_CPUS__);
+} CPU_MAP __section_maps_btf;
```
Don't bother with the `CPU_MAP`-type macros, they haven't been necessary for years. This can simply be called `cpu_map` here and everywhere in the code. This also gives the map a stable name in the Collection(Spec), so changing MapSpec.MaxEntries becomes less error-prone.

ti-mo commented on bpf/lib/maps.h:
```diff
@@ -284,3 +285,13 @@ tail_call_internal(struct __ctx_buff *ctx, const __u32 index, __s8 *ext_err)
+    __type(key, __u32);
+    __type(value, struct bpf_cpumap_val);
+    __uint(pinning, LIBBPF_PIN_BY_NAME);
+    __uint(max_entries, __NR_ONLINE_CPUS__);
```
Instead of leaning on the preprocessor for setting maxentries, I suggest we add a small case to ebpf-go to automatically set it when it's left at its default 0 value as well as clamping it to possible CPUs, like we do for PerfEventArray: https://github.com/cilium/ebpf/blob/9895aae6467e9f677846e5b1407893133dbbbfc9/map.go#L125.

Also, I'm not sure if limiting it to online CPUs makes sense, since the list can have holes and that would mean some CPUs get cordoned off, while others are wrongly included. For example, on a 16-core machine, if CPU 0 and 1 go offline, the cpumap would have 14 entries, wrongly including CPU 0 and 1 and leaving online CPUs 14 and 15 unutilized?

That said, I'm not sure if bpf can deal with offline CPUs here at all, since https://docs.kernel.org/bpf/map_cpumap.html only mentions bpf code using `bpf_num_possible_cpus()` and setting `libbpf_num_possible_cpus()` during map creation.

learnitall replied:
Hey @ti-mo, I gave implementing setting the max entries a shot, see https://github.com/cilium/ebpf/pull/1627. It sets the max entries to the number of possible CPUs on the node. I'm thinking we can change the logic used to select a CPU index to account for online CPUs, rather than setting the max entries to online CPUs.

ti-mo commented on bpf/lib/maps.h:
```diff
@@ -284,3 +285,13 @@ tail_call_internal(struct __ctx_buff *ctx, const __u32 index, __s8 *ext_err)
+    __uint(type, BPF_MAP_TYPE_CPUMAP);
+    __type(key, __u32);
+    __type(value, struct bpf_cpumap_val);
+    __uint(pinning, LIBBPF_PIN_BY_NAME);
```
This is dangerous if CPU maps ever contain program pointers for multi-xdp. Loading multiple ELFs (1 per xdp device) will start clobbering entries in this shared map and lead to a split-brain where packets entering one interface will end up executing programs belonging to another.

See my other comments, there's little value in pinning this, and not doing so will save us 500 lines of boilerplate.

learnitall replied:
To make sure I understand, you're referencing a use case where we use the CPU MAP in the future to implement multi-xdp for different use cases, and those use cases collide with one another since the map is shared?

ti-mo commented on bpf/node_config.h:
```diff
@@ -336,3 +336,7 @@ return false;
 
 #define CALLS_MAP test_cilium_calls_65535
+
+#ifdef ENABLE_CPU_MAP
```
No need for this, see comment above.

ti-mo commented on pkg/datapath/linux/ipsecrps/ipsec_rps_linux.go:
```diff
@@ -0,0 +1,142 @@
+	defines := make(dpcfgdef.Map)
+
+	if cfg.Enabled() {
+		defines["ENABLE_IPSEC_RPS"] = "1"
```
Feel free to make this a `DECLARE_CONFIG(bool, enable_ipsec_rps, "...")` at the top of bpf_xdp.c and specify it as a constant during XDP loading. I'll have more to share soon, but this will automatically include it in datapath config scaffolding codegen that we'll migrate all configurables to over time.

ti-mo commented on bpf/bpf_host.c:
```diff
@@ -1075,6 +1076,28 @@ static __always_inline int handle_l2_announcement(struct __ctx_buff *ctx)
 };
 #endif
 
+#if defined(ENABLE_IPSEC) && defined(ENABLE_IPSEC_RPS)
```
Is there a benefit to omitting this in the preprocessor? Can the call site be guarded with `if CONFIG(enable_ipsec_rps) { ... }` instead? (see my other comments)

ti-mo commented on bpf/bpf_xdp.c:
```diff
@@ -36,6 +37,10 @@
 #include "lib/events.h"
 #include "lib/nodeport.h"
 
+#if defined(ENABLE_IPSEC) && defined(ENABLE_IPSEC_RPS)
```
Can this be included unconditionally?

ti-mo commented on bpf/bpf_xdp.c:
```diff
@@ -234,15 +261,35 @@ static __always_inline int check_v4(struct __ctx_buff *ctx)
+		return CTX_ACT_DROP;
 #endif /* CIDR4_FILTER */
+
+#if defined(ENABLE_IPSEC_RPS)
```
Try using a natural `if CONFIG(enable_ipsec_rps) { ...` statement here, see other comments.

ti-mo reviewed (changes requested):
I left a few comments, I think this is way way too much code for what it does, and the concept of dealing with offline CPUs I think needs a bit of clarification. Please reach out if you have any questions regarding the new config mechanism. It's not quite ready for prime time, but it's usable, so please give it a try!
